### DISCUSSION TITLE
How to make EB2::build faster

### DISCUSSION THREAD

**Discussion by FR13ndSDP** [AMReX-Codes/amrex#4130] [Q&A] [answered]
I'm trying to use AMReX to do wind turbine simulations. Now I have STL format geometry and want it to rotate at every timestep. It's a complex geometry and the mesh quantity would be extremely large for its representation. The problem I encounter is that building EB information is very slow, let alone every timestep I need to rebuild it. So I'm wondering if there is any trick to make `EB2::build` faster.

**Comment by rmrsk** [comment]
If you're not doing it already, you can use eb2.num_coarsen_opt to improve load balancing. 

How many triangles in your STL?

**Reply by FR13ndSDP** [reply]
Hi Robert, thanks for your advice and I'll definitely try this. There are 601858 triangles in my STL. The difficulty is that it moves, so I have to build EB information at every time step. Maybe this is stupid, but I don't have a better idea.

**Reply by rmrsk** [reply]
Ok, can you share the STL? I'd like to try it

**Reply by FR13ndSDP** [reply]
Sure, here is the link: [Blade-STL](https://hkustconnect-my.sharepoint.com/:u:/g/personal/dsongak_connect_ust_hk/EVfG6aVf8VZMmCtbd5eWns0B7gPzMmOBAkHrVADpKRdN3Q?e=Uu7k3M)

**Reply by FR13ndSDP** [reply]
I've reduced the number of triangles with some tools.

**Comment by rmrsk** [comment]
Good news and bad news, I guess. 

The good news is that we have code (https://github.com/rmrsk/EBGeometry) for turning STLs into distance fields, and this is done by means of hierarchical structures that perform _way_ better than AMReX's native STL reader once you start hitting 10k triangles and beyond.

The bad news is that 1) we haven't (yet) put this on the GPU, and 2) the surface must be manifold (watertight, no self-intersections, and coherently oriented faces). Dirty CAD files, in general, don't have very robust distance fields, so we haven't made much of an effort to support them.

**Reply by FR13ndSDP** [reply]
Many thanks! Do you think, for this geometry, it is possible to make it have the good properties you have mentioned? Your code seems very promising for handling complex geometries.

**Reply by rmrsk** [reply]
That's probably just a matter of repairing your STL file. I normally use meshlab or meshinspector foe that

**Reply by FR13ndSDP** [reply]
Great! I'll try it.

**Comment by WeiqunZhang** [comment, marked as answer]
I have submitted a draft PR (#4140) trying to address the performance issue. The code has not been thoroughly tested. But please feel free to give it a try and report any issues.

**Reply by rmrsk** [reply]
Glad to see that I kept you up this weekend - the performance increase is as expected.

The BVH build and traversal looks similar to top-down construction and recursionless traversal from EBGeometry. Perhaps an acknowledgement in the PR text would be in order?

**Reply by FR13ndSDP** [reply]
Great! For my case I found code from @rmrsk works well and significant speedup was observed. I can't wait to try this on code compiled for GPUs.

**Reply by WeiqunZhang** [reply]
> Glad to see that I kept you up this weekend

That's the life of a programmer. :smile:
